store/cachekv/internal/btree.go

```diff
@@ -0,0 +1,85 @@
+package internal
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/tidwall/btree"
+)
+
+const (
+	// The approximate number of items and children per B-tree node. Tuned with benchmarks.
+	// copied from memdb.
+	bTreeDegree = 32
+)
+
+var errKeyEmpty = errors.New("key cannot be empty")
+
+// BTree implements the sorted cache for cachekv store,
+// we don't use MemDB here because cachekv is used extensively in sdk core path,
+// we need it to be as fast as possible, while `MemDB` is mainly used as a mocking db in unit tests.
```

Comment:
funny thing about memdb is it use to be a simple array cause it was meant for testing then while I was working on tendermint I discovered sdk uses it in the a production path. This may be where the bottle neck was introduced since we migrated to btree.